src/chainbin/nnet3-chain-get-egs.cc

```diff
@@ -30,50 +30,68 @@
                               const MatrixBase<BaseFloat> *ivector_feats,
                               NnetChainExample *eg) {
-  // try to get a representative iVector every ivector_period frames
-  // in an eg.
   KALDI_ASSERT(ivector_feats->NumRows() > 0);
-  // num_ivectors_in_left_context is the num of ivectors for frames whose t < 0.
-  // It is used to compute the initial value of closest_frame
-  // num_ivectors is the num of ivectors for frames whose t >= 0.
-  // Both of them are computed according to how Round descriptor works, which
-  // basically returns floor(t / <t-modulus>) * <t-modulus>.
-  int32 num_ivectors_in_left_context =
-      std::ceil(1.0 * left_context / ivector_period); 
-  int32 num_ivectors = num_ivectors_in_left_context +
-      (frames_per_eg + right_context - 1) / ivector_period + 1;
-  Matrix<BaseFloat> ivectors(num_ivectors, ivector_feats->NumCols());
-  // closest_frame is the frame at whole multiples of ivector_period
-  int32 closest_frame = t - num_ivectors_in_left_context * ivector_period;
-  for (int32 n = 0; n < num_ivectors; n++) {
-    if (closest_frame < 0)
-      closest_frame = 0;
-    else if (closest_frame >= ivector_feats->NumRows())
+  if (ivector_period == 0) {
+    // single ivector case.
+    // if applicable, add the iVector feature.
+    // try to get closest frame to middle of window to get
+    // a representative iVector.
+    int32 closest_frame = t + frames_per_eg / 2;
+    KALDI_ASSERT(ivector_feats->NumRows() > 0);
+    if (closest_frame >= ivector_feats->NumRows())
       closest_frame = ivector_feats->NumRows() - 1;
-    ivectors.Row(n).CopyFromVec(ivector_feats->Row(closest_frame));
-    closest_frame += ivector_period;
+    Matrix<BaseFloat> ivector(1, ivector_feats->NumCols());
+    ivector.Row(0).CopyFromVec(ivector_feats->Row(closest_frame));
+    NnetIo ivector_io("ivector", 0, ivector);
+    eg->inputs[1].Swap(&ivector_io);
+  } else {
+    // multiple ivectors case.
+    // try to get a representative iVector every ivector_period frames
+    // in an eg.
+    // num_ivectors_in_left_context is the num of ivectors for frames whose
+    // t < 0. It is used to compute the initial value of closest_frame
+    // num_ivectors is the num of ivectors for frames whose t >= 0.
+    // Both of them are computed according to how Round descriptor works, which
+    // basically returns floor(t / <t-modulus>) * <t-modulus>.
+    int32 num_ivectors_in_left_context =
+        -1 * DivideRoundingDown(-left_context, ivector_period); 
+    int32 num_ivectors = num_ivectors_in_left_context + 1 +
+        DivideRoundingDown(frames_per_eg + right_context - 1, ivector_period);
+    Matrix<BaseFloat> ivectors(num_ivectors, ivector_feats->NumCols());
+    // closest_frame is the frame at whole multiples of ivector_period
+    int32 closest_frame = t - num_ivectors_in_left_context * ivector_period;
+    for (int32 n = 0; n < num_ivectors; n++) {
+      int32 ivector_frame = closest_frame;
```

Comment:
@danpovey Did you mean I need to have a different variable like ivector_frame here, and clip ivector_frame if it is out of the range, while keeping closest_frame always at whole multiples of the period?

Yiming